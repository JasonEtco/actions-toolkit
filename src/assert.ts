import { Exit } from './exit'

export class Assert {
  private exit: Exit

  constructor (exit: Exit) {
    this.exit = exit
  }

  /**
   * Assert condition is truthy and exit with "failed" status if not
   */
  public isTruthy(condition: any, msg?: string): asserts condition {
    if (!condition) {
      this.exit.failure(msg ?? `Assertion for value ${condition} is not truthy`)
    }
  }
}
