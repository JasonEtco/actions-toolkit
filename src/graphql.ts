import { graphql } from '@octokit/graphql'

export const withDefaults = (token: string) => graphql.defaults({
  headers: { authorization: `token ${token}` }
})
