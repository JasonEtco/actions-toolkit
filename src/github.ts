import { graphql } from '@octokit/graphql'
import Octokit from '@octokit/rest'
import { withDefaults } from './graphql'

export class GitHub extends Octokit {
  public graphql: typeof graphql

  constructor (token: string) {
    super({ auth: `token ${token}` })
    this.graphql = withDefaults(token)
  }
}
