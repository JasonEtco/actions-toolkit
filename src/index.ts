import execa, { Options as ExecaOptions } from 'execa'
import fs from 'fs'
import yaml from 'js-yaml'
import minimist, { ParsedArgs } from 'minimist'
import path from 'path'
import { LoggerFunc, Signale } from 'signale'
import { Context } from './context'
import { Exit } from './exit'
import { GitHub } from './github'
import { Store } from './store'

export interface ToolkitOptions {
  event?: string | string[],
  logger?: Signale
}

export class Toolkit {
  public context: Context

  /**
   * A key/value store for arbitrary data that can be accessed across actions in a workflow
   */
  public store: Store

  /**
   * Path to a clone of the repository
   */
  public workspace: string

  /**
   * GitHub API token
   */
  public token: string

  /**
   * An object of the parsed arguments passed to your action
   */
  public arguments: ParsedArgs

  /**
   * An Octokit SDK client authenticated for this repository. See https://octokit.github.io/rest.js for the API.
   *
   * ```js
   * const newIssue = await tools.github.issues.create(context.repo({
   *   title: 'New issue!',
   *   body: 'Hello Universe!'
   * }))
   * ```
   */
  public github: GitHub

  public opts: ToolkitOptions

  /**
   * A collection of methods used to stop an action while it's being run
   */
  public exit: Exit

  /**
   * A general-purpose logger. An instance of [Signale](https://github.com/klaussinani/signale)
   */
  public log: Signale & LoggerFunc

  constructor (opts: ToolkitOptions = {}) {
    this.opts = opts

    // Disable the underline to prevent extra white space in the Actions log output
    this.log = this.wrapLogger(opts.logger)

    // Print a console warning for missing environment variables
    this.warnForMissingEnvVars()

    this.exit = new Exit()
    this.context = new Context()
    this.workspace = process.env.GITHUB_WORKSPACE as string
    this.token = process.env.GITHUB_TOKEN as string
    this.github = new GitHub(this.token)
    this.arguments = minimist(process.argv.slice(2))
    this.store = new Store(this.context.workflow, this.workspace)
    this.checkAllowedEvents()
  }

  /**
   * Gets the contents file in your project's workspace
   *
   * ```js
   * const myFile = tools.getFile('README.md')
   * ```
   *
   * @param filename - Name of the file
   * @param encoding - Encoding (usually utf8)
   */
  public getFile (filename: string, encoding = 'utf8') {
    const pathToFile = path.join(this.workspace, filename)
    if (!fs.existsSync(pathToFile)) throw new Error(`File ${filename} could not be found in your project's workspace.`)
    return fs.readFileSync(pathToFile, encoding)
  }

  /**
   * Get the package.json file in the project root
   *
   * ```js
   * const pkg = toolkit.getPackageJSON()
   * ```
   */
  public getPackageJSON (): object {
    const pathToPackage = path.join(this.workspace, 'package.json')
    if (!fs.existsSync(pathToPackage)) throw new Error('package.json could not be found in your project\'s root.')
    return require(pathToPackage)
  }

  /**
   * Get the configuration settings for this action in the project workspace.
   *
   * @param key - If this is a string like `.myfilerc` it will look for that file.
   * If it's a YAML file, it will parse that file as a JSON object. Otherwise, it will
   * return the value of the property in the `package.json` file of the project.
   *
   * @example This method can be used in three different ways:
   *
   * ```js
   * // Get the .rc file
   * const cfg = toolkit.config('.myactionrc')
   *
   * // Get the YAML file
   * const cfg = toolkit.config('myaction.yml')
   *
   * // Get the property in package.json
   * const cfg = toolkit.config('myaction')
   * ```
   */
  public config (key: string): object {
    if (/\..+rc/.test(key)) {
      // It's a file like .npmrc or .eslintrc!
      return JSON.parse(this.getFile(key))
    } else if (key.endsWith('.yml') || key.endsWith('.yaml')) {
      // It's a YAML file! Gotta serialize it!
      return yaml.safeLoad(this.getFile(key))
    } else {
      // It's a regular object key in the package.json
      const pkg = this.getPackageJSON() as any
      return pkg[key]
    }
  }

  /**
   * Run a CLI command in the workspace. This runs [execa](https://github.com/sindresorhus/execa)
   * under the hood so check there for the full options.
   *
   * @param command - Command to run
   * @param args - Argument (this can be a string or multiple arguments in an array)
   * @param cwd - Directory to run the command in
   * @param [opts] - Options to pass to the execa function
   */
  public async runInWorkspace (command: string, args?: string[] | string, opts?: ExecaOptions) {
    if (typeof args === 'string') args = [args]
    return execa(command, args, { cwd: this.workspace, ...opts })
  }

  /**
   * Create a scoped logger. Also an instance of [Signale](https://github.com/klaussinani/signale)
   * This will prefix any logs with the name provided by @loggerScope
   * @param loggerScope - Name of the scope
   * @param logger - Provide a customized Signale logger
   */
  public createScopedLogger (loggerScope: string, logger?: Signale) {
    if (!logger) logger = new Signale({ config: { underlineLabel: false } })
    const loggerWithScope = logger.scope(loggerScope)
    const fn = loggerWithScope.info.bind(loggerWithScope)
    return Object.assign(fn, loggerWithScope)
  }

  /**
   * Returns true if this event is allowed
   */
  private eventIsAllowed (event: string) {
    const [eventName, action] = event.split('.')

    if (action) {
      return eventName === this.context.event && this.context.payload.action === action
    }

    return eventName === this.context.event
  }

  private checkAllowedEvents () {
    const { event } = this.opts
    if (!event) return

    const passed = Array.isArray(event)
      ? event.some(e => this.eventIsAllowed(e))
      : this.eventIsAllowed(event)

    if (!passed) {
      const actionStr = this.context.payload.action ? `.${this.context.payload.action}` : ''
      this.log.error(`Event \`${this.context.event}${actionStr}\` is not supported by this action.`)
      this.exit.neutral()
    }
  }

  /**
   * Wrap a Signale logger so that its a callable class
   */
  private wrapLogger (logger?: Signale) {
    if (!logger) logger = new Signale({ config: { underlineLabel: false } })
    const fn = logger.info.bind(logger)
    return Object.assign(fn, logger)
  }

  /**
   * Log warnings to the console for missing environment variables
   */
  private warnForMissingEnvVars () {
    const requiredEnvVars = [
      'HOME',
      'GITHUB_WORKFLOW',
      'GITHUB_ACTION',
      'GITHUB_ACTOR',
      'GITHUB_REPOSITORY',
      'GITHUB_EVENT_NAME',
      'GITHUB_EVENT_PATH',
      'GITHUB_WORKSPACE',
      'GITHUB_SHA',
      'GITHUB_REF',
      'GITHUB_TOKEN'
    ]

    const requiredButMissing = requiredEnvVars.filter(key => !process.env.hasOwnProperty(key))
    if (requiredButMissing.length > 0) {
      // This isn't being run inside of a GitHub Action environment!
      const list = requiredButMissing.map(key => `- ${key}`).join('\n')
      const warning = `There are environment variables missing from this runtime, but would be present on GitHub.\n${list}`
      this.log.warn(warning)
    }
  }
}
