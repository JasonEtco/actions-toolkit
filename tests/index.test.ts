import nock from 'nock'
import path from 'path'
import { Signale } from 'signale'
import { Toolkit } from '../src'
import { NeutralCode } from '../src/exit'

describe('Toolkit', () => {
  let toolkit: Toolkit

  beforeEach(() => {
    toolkit = new Toolkit({ logger: new Signale({ disabled: true }) })
  })

  describe('#github', () => {
    it('returns a GitHub client', () => {
      expect(toolkit.github).toBeInstanceOf(Object)
    })

    it('returns a GraphQL function on `.graphql`', async () => {
      expect(toolkit.github.graphql).toBeInstanceOf(Function)

      const scoped = nock('https://api.github.com')
        .post('/graphql').reply(200, { data: { errors: [] } })

      await toolkit.github.graphql('query { }')
      expect(scoped.isDone()).toBe(true)
    })
  })

  describe('#getFile', () => {
    it('gets the contents of a file', () => {
      const actual = toolkit.getFile('README.md')
      expect(actual).toMatchSnapshot()
    })

    it('gets the contents of a file with custom encoding', () => {
      const actual = toolkit.getFile('README.md', 'base64')
      expect(actual).toMatchSnapshot()
    })

    it('throws if the file could not be found', () => {
      const actual = () => toolkit.getFile('DONTREADME.md')
      expect(actual).toThrowErrorMatchingSnapshot()
    })
  })

  describe('#getPackageJSON', () => {
    it('returns the package.json file as a JSON object', () => {
      const actual = toolkit.getPackageJSON()
      expect(actual).toMatchSnapshot()
    })

    it('throws if the package.json file could not be found', () => {
      toolkit.workspace = path.join(__dirname, 'fixtures', 'workspaces', 'no-package-json')
      const actual = () => toolkit.getPackageJSON()
      expect(actual).toThrowErrorMatchingSnapshot()
    })
  })

  describe('#config', () => {
    it('returns a property in the package.json', () => {
      const actual = toolkit.config('action')
      expect(actual).toEqual({ foo: true })
    })

    it('returns a parsed YAML file', () => {
      const actual = toolkit.config('action.yml')
      expect(actual).toEqual({ foo: true })
    })

    it('returns a .rc file as JSON', () => {
      const actual = toolkit.config('.actionrc')
      expect(actual).toEqual({ foo: true })
    })
  })

  describe('#runInWorkspace', () => {
    it('runs the command in the workspace', async () => {
      const result = await toolkit.runInWorkspace('echo', 'hello')
      expect(result).toMatchSnapshot()
      expect(result.stdout).toBe('hello')
    })

    it('runs the command in the workspace with some options', async () => {
      const result = await toolkit.runInWorkspace('throw', undefined, { reject: false })
      expect(result).toMatchSnapshot()
    })
  })

  describe('#wrapLogger', () => {
    it('wraps the provided logger and allows for a callable class', () => {
      const logger = new Signale() as jest.Mocked<Signale>
      logger.info = jest.fn()
      const twolkit = new Toolkit({ logger })

      twolkit.log('Hello!')
      twolkit.log.info('Hi!')

      expect(logger.info).toHaveBeenCalledTimes(2)
      expect(logger.info).toHaveBeenCalledWith('Hello!')
      expect(logger.info).toHaveBeenCalledWith('Hi!')
    })
  })
})

describe('Toolkit#constructor', () => {
  let logger: jest.Mocked<Signale>
  let exit: (code?: number) => never

  beforeEach(() => {
    logger = new Signale() as jest.Mocked<Signale>
    logger.error = jest.fn()
    logger.warn = jest.fn()

    exit = global.process.exit
    const p = global.process as any
    p.exit = jest.fn()
  })

  it('exits if the event is not allowed with an array of events', () => {
    // tslint:disable-next-line:no-unused-expression
    new Toolkit({ logger, event: ['pull_request'] })
    expect(process.exit).toHaveBeenCalledWith(NeutralCode)
    expect(logger.error.mock.calls).toMatchSnapshot()
  })

  it('does not exit if the event is one of the allowed with an array of events', () => {
    // tslint:disable-next-line:no-unused-expression
    new Toolkit({ logger, event: ['pull_request', 'issues'] })
    expect(process.exit).not.toHaveBeenCalled()
    expect(logger.error).not.toHaveBeenCalled()
  })

  it('exits if the event is not allowed with a single event', () => {
    // tslint:disable-next-line:no-unused-expression
    new Toolkit({ logger, event: 'pull_request' })
    expect(process.exit).toHaveBeenCalledWith(NeutralCode)
    expect(logger.error.mock.calls).toMatchSnapshot()
  })

  it('exits if the event is not allowed with an array of events with actions', () => {
    // tslint:disable-next-line:no-unused-expression
    new Toolkit({ logger, event: ['pull_request.opened'] })
    expect(process.exit).toHaveBeenCalledWith(NeutralCode)
    expect(logger.error.mock.calls).toMatchSnapshot()
  })

  it('exits if the event is not allowed with a single event with an action', () => {
    // tslint:disable-next-line:no-unused-expression
    new Toolkit({ logger, event: 'pull_request.opened' })
    expect(process.exit).toHaveBeenCalledWith(NeutralCode)
    expect(logger.error.mock.calls).toMatchSnapshot()
  })

  it('logs the expected string with missing env vars', () => {
    delete process.env.HOME
    new Toolkit({ logger }) // tslint:disable-line:no-unused-expression
    expect(logger.warn.mock.calls).toMatchSnapshot()
  })

  it('sets a named logger', () => {
    const tools = new Toolkit({ logger }) // tslint:disable-line:no-unused-expression
    const fooLogger = {
      success: jest.fn().mockImplementation(tools.createScopedLogger('foo').success)
    }
    fooLogger.success('hi there from foo logger')
    expect(fooLogger.success.mock.calls).toMatchSnapshot()
  })

  afterEach(() => {
    global.process.exit = exit
  })
})
