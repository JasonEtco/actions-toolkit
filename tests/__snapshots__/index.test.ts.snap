// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolkit #getFile gets the contents of a file 1`] = `"I'm writing this on a plane ✈️"`;

exports[`Toolkit #getFile gets the contents of a file with custom encoding 1`] = `"SSdtIHdyaXRpbmcgdGhpcyBvbiBhIHBsYW5lIOKciO+4jw=="`;

exports[`Toolkit #getFile throws if the file could not be found 1`] = `"File DONTREADME.md could not be found in your project's workspace."`;

exports[`Toolkit #getPackageJSON returns the package.json file as a JSON object 1`] = `
Object {
  "action": Object {
    "foo": true,
  },
}
`;

exports[`Toolkit #getPackageJSON throws if the package.json file could not be found 1`] = `"package.json could not be found in your project's root."`;

exports[`Toolkit #runInWorkspace runs the command in the workspace 1`] = `
Object {
  "cmd": "echo hello",
  "code": 0,
  "failed": false,
  "killed": false,
  "signal": null,
  "stderr": "",
  "stdout": "hello",
  "timedOut": false,
}
`;

exports[`Toolkit #runInWorkspace runs the command in the workspace with some options 1`] = `[Error: spawn throw ENOENT]`;

exports[`Toolkit#constructor exits if the event is not allowed with a single event 1`] = `
Array [
  Array [
    "Event \`issues.opened\` is not supported by this action.",
  ],
]
`;

exports[`Toolkit#constructor exits if the event is not allowed with a single event with an action 1`] = `
Array [
  Array [
    "Event \`issues.opened\` is not supported by this action.",
  ],
]
`;

exports[`Toolkit#constructor exits if the event is not allowed with an array of events 1`] = `
Array [
  Array [
    "Event \`issues.opened\` is not supported by this action.",
  ],
]
`;

exports[`Toolkit#constructor exits if the event is not allowed with an array of events with actions 1`] = `
Array [
  Array [
    "Event \`issues.opened\` is not supported by this action.",
  ],
]
`;

exports[`Toolkit#constructor logs the expected string with missing env vars 1`] = `
Array [
  Array [
    "There are environment variables missing from this runtime, but would be present on GitHub.
- HOME",
  ],
]
`;

exports[`Toolkit#constructor sets a named logger 1`] = `
Array [
  Array [
    "hi there from foo logger",
  ],
]
`;
